### PR TITLE
handler negative size/from paramter in search request

### PR DESCRIPTION
**Pull request by makeyang**
throw exception when user set negative size or negative from paramter in search request, which is describe in : https://github.com/elastic/elasticsearch/issues/22530

### CODE REVIEW

**Comment by gaganis** [comment]
Hello @makeyang, 

Just a quick note on your PR. I see that that with this code you make it illegal to request size of "-1".
There might be clients that knowingly or unknowingly are already using this and their code would break if this behaviour is changed. I would have kept it to return the default size of 10 which is clearly documented here [](https://www.elastic.co/guide/en/elasticsearch/reference/5.1/search-request-from-size.html)

What I mean by unknowingly is that many people are using -1 as the uninitialized value for an int variable so it is possible in some cases this could leak through to the API. 

Giorgos

**Comment by makeyang** [comment]
@gaganis actually I tought about it much when I try to fix this. and here is my decision: u just can't let -1 as the default value while throw exception for -2. this is much much ambiguous for more users than bother those who already known that -1 is default value.
we can mark breaking change if it is necessary. but the bottom line is: we really should make the right things rather than make things right.

**Comment by jasontedor** [comment]
An idiom within Elasticsearch is that -1 means disabled or default or unset (depending on the context). This needs to remain as-is for this API too.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 1]
```diff
@@ -126,9 +126,9 @@ public static HighlightBuilder highlight() {
 
     private QueryBuilder postQueryBuilder;
 
-    private int from = -1;
+    private Integer from = null;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: I think that this change that introduces boxing should be reverted. This means that we need a value to represent unset, and -1 is the idiom within Elasticsearch for that.

File: core/src/main/java/org/elasticsearch/rest/RestRequest.java [hunk 2]
```diff
@@ -186,6 +186,18 @@ public int paramAsInt(String key, int defaultValue) {
         }
     }
 
+    public Integer paramAsInteger(String key, Integer defaultValue) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/rest/RestRequest.java]: This method isn't needed since I think the boxing change should be reverted.

**Comment by makeyang** [comment]
@jasontedor & @gaganis I just checked ES reference and definitive gudie, there is nowhere mentioned this so-called idiom. it's just the code convention.
I really don't think that is the right thing. 
but if u guys insist on, I'll modified code.
let me know u guys thouths.

File: core/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java [hunk 3]
```diff
@@ -130,15 +130,12 @@ private static void parseSearchSource(final SearchSourceBuilder searchSourceBuil
             searchSourceBuilder.query(queryBuilder);
         }
 
-        int from = request.paramAsInt("from", -1);
-        if (from != -1) {
-            searchSourceBuilder.from(from);
+        if (request.hasParam("from")) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java]: I would be okay with rejecting anything negative here, on the REST layer. We still need a way to represent unset internally. Using `null` introduces boxing, we should avoid that. Let's continue to use -1, that's the idiom.

However, I would like to solicit feedback from others before you proceed with rejecting anything negative including -1 here. What do you think @javanna?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left some comments.

File: core/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java (hunk 3 above)
Comment (clintongormley) [inline core/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java]: Accepting -1 over the rest layer has never been documented, and it surprised me that it works. I'd be ok with rejecting it

Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java]: given our direction (REST client vs transport client) I think we are fine rejecting `-1` on the REST layer. But how about keeping the `int`, using `-1` as default, and rejecting negative values in the setters? That's how we changed some of the request validation as part of the search refactoring too.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java]: @javanna Yeah, that's exactly what I have in mind.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java]: Thank you for commenting @clintongormley and @javanna.

**Comment by gaganis** [comment]
This has become a pretty interesting discussion.

It seems the discussion has revolved around two distinct points:

#### Data type of the Internal implementation 

Whether to use int with a default value of -1 for the endpoint implementing classes. On my original comment I was not thinking about that but after reading the rest of the comments I would also agree to keep the int data type and the -1 as the default value internally.

#### What to do with the undocumented behavior or the REST endpoint accepting -1

This was the focus of my initial comment. This behavior despite being undocumented was preexisting and the change could potentially break existing clients. The consensus seems to be that this is not correct and should be fixed in line with @makeyang's reasoning. 

I have done some more research and I have found the very visible and extensive section in the online documentation for changes that have the potential to break existing code and behavior. BTW That is pretty cool! 

#### In summary 

Taking into account all the comments and the availability of the breaking section in the docs, @makeyang my thoughts on this is that we should keep the int and -1 default value internally and accept rejecting -1 during the validation per your implementation.

Thanks @makeyang for the PR and @jasontedor, @clintongormley and @javanna for the comments!

They were all very informative!

**Comment by makeyang** [comment]
@jasontedor can you help to review this? I have modified code according to your comments.

**Comment by makeyang** [comment]
@jasontedor  @gaganis 
actually I rejecting negative values in the setters of SearchSourceBuilder.

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 4]
```diff
@@ -304,6 +304,9 @@ public QueryBuilder postFilter() {
      * From index to start the search from. Defaults to <tt>0</tt>.
      */
     public SearchSourceBuilder from(int from) {
+        if (from < 0) {
+            throw new IllegalArgumentException("from must be no negative but was [" + from + "]");
```
Comment (bumbu) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: A better wording for the error is: `[from] can't be negative`
Changing that will require updating tests

File: core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 5]
```diff
@@ -319,6 +322,9 @@ public int from() {
      * The number of search hits to return. Defaults to <tt>10</tt>.
      */
     public SearchSourceBuilder size(int size) {
+        if (size < 0) {
+            throw new IllegalArgumentException("size must be no negative but was [" + size + "]");
```
Comment (bumbu) [inline core/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java]: A better wording for the error is: `[size] can't be negative`
Changing that will require updating tests

**Review by bumbu — CHANGES_REQUESTED** [review]
Updates on errors wording

File: core/src/test/java/org/elasticsearch/search/builder/SearchSourceBuilderTests.java [hunk 6]
```diff
@@ -366,6 +366,26 @@ public void testParseIndicesBoost() throws IOException {
         }
     }
 
+    public void testInvalidFormAndSize() throws IOException {
```
Comment (olcbean) [inline core/src/test/java/org/elasticsearch/search/builder/SearchSourceBuilderTests.java]: A tiny typo in the test name : should be `From` instead of `Form`

**Comment by jasontedor** [comment]
@makeyang Are you interested in finishing this pull request?

**Comment by makeyang** [comment]
@jasontedor  I'll update this later this month.

**Comment by colings86** [comment]
No further feedback, closing. Please reopen if you have progress on this
